File: backend/app/modules/inventory/depletion/units.py
"""Canonical unit allowlist — the single Python source of truth.

Mirrors the DB CHECK constraints on recipe_ingredients.unit / modifier_ingredients.unit
/ unit_conversions.from_unit/to_unit (migrations 0014, 0016). API and UI validators
import from here so the allowlist is defined in exactly one place in Python.

'oz' is intentionally absent — use 'oz_weight' (mass) or 'fl_oz' (volume).
"""

from __future__ import annotations

_WEIGHT: tuple[str, ...] = ("g", "kg", "oz_weight", "lb")
_VOLUME: tuple[str, ...] = ("ml", "L", "fl_oz", "cup", "tsp", "tbsp")
_COUNT: tuple[str, ...] = ("ea", "dozen")

# unit -> dimension. A conversion is only ever defined within one dimension at
# the global tier; cross-dimension (e.g. cup -> g) requires a per-item density
# override (see unit_conversions item tier).
DIMENSION_OF: dict[str, str] = {
    **{u: "weight" for u in _WEIGHT},
    **{u: "volume" for u in _VOLUME},
    **{u: "count" for u in _COUNT},
}

CANONICAL_UNITS: frozenset[str] = frozenset(DIMENSION_OF)


def is_canonical(unit: str) -> bool:
    """True if ``unit`` is in the canonical allowlist."""
    return unit in CANONICAL_UNITS

File: frontend/src/api/units.ts
// Canonical unit allowlist — PINNED to the backend source of truth:
//   backend/app/modules/inventory/depletion/units.py  (CANONICAL_UNITS)
//   + the DB CHECK constraints (migrations 0014 / 0016).
//
// Drift is NOT a judgment call: the backend test
// `backend/tests/test_frontend_units_sync.py` parses THIS file and FAILS CI if its
// set ever diverges from units.py. The picker must offer exactly what the API
// accepts — otherwise an operator picks a unit the UI offered and the API rejects it
// with a confusing 400 (client validation is UX; the DB CHECK + API 400 is enforcement).
//
// 'oz' is intentionally absent — use 'oz_weight' (mass) or 'fl_oz' (volume).

export type Dimension = 'weight' | 'volume' | 'count';

export const CANONICAL_UNITS_BY_DIMENSION: Record<Dimension, readonly string[]> = {
  weight: ['g', 'kg', 'oz_weight', 'lb'],
  volume: ['ml', 'L', 'fl_oz', 'cup', 'tsp', 'tbsp'],
  count: ['ea', 'dozen'],
};

export const CANONICAL_UNITS: readonly string[] = [
  ...CANONICAL_UNITS_BY_DIMENSION.weight,
  ...CANONICAL_UNITS_BY_DIMENSION.volume,
  ...CANONICAL_UNITS_BY_DIMENSION.count,
];

export function isCanonical(unit: string): boolean {
  return CANONICAL_UNITS.includes(unit);
}

/** Dimension of a canonical unit; null for non-canonical strings (CS, SAC, ...). */
export function dimensionOf(unit: string): Dimension | null {
  for (const d of Object.keys(CANONICAL_UNITS_BY_DIMENSION) as Dimension[]) {
    if (CANONICAL_UNITS_BY_DIMENSION[d].includes(unit)) return d;
  }
  return null;
}
